### DISCUSSION TITLE
Support for 10-10-10-2 formats? (Or: how to pass 10-bit raw images to ImageMagick?)

### DISCUSSION THREAD

**Discussion by Oblomov** [ImageMagick/ImageMagick#6030] [Help]
### ImageMagick version

7.1.0-58 Q16-HDR

### Operating system, version and so on

Linux, Debian unstable

### Description

Hello all,

I'm working with some multi-spectral/hyper-spectral images, and I would like to produce some standard RGB images or animations from the raw data (no processing). The source images have a depth of 10 bits per sample. The way I'm currently handling this is by upconverting each channel to a bit depth of 16 and then producing MIFF images with this, that get piped to ImageMagick (or stored for later processing or whatever).

Considering that there are image formats that pack 3 10-bit channels (plus an optional 2-bit alpha channel) into a single 32-bit per pixel (taking 1/3rd of the memory or on-disk storage, or even less in case of simple alpha), I was wondering if ImageMagick supported this kind of representation at the MIFF level, so that I could produce more compact data from my scripts and reduce transfer times when going through the network. I understand that internally ImageMagick would still end up representing the data as 16-bit per sample, but the intermediate requirements would help me reduce the cost of some parts of the processing.

**Comment by snibgo** [comment]
MIFF files are fairly simple image files proprietary to ImageMagick. It has some options for compression. I don't think there is much benefit in extending this to unusual formats such as 10:10:10:2.

If 10:10:10:2 is a public format, with published specification, then a coder could be written to read and write files of that format.

**Comment by Oblomov** [comment]
Thanks for your reply. RLE could definitely be another way to reduce the output size. If there is also some way to specify that the maximum value is 1023 instead of 65535, it would simply my work on the production side even more.

Concerning file formats, BMP is an example of file format that supports 10:10:10:2 data. For raw, many GPUs support surfaces/textures with raw data in this format. This is similar to the 5:6:5  format already supported by ImageMagick. There are actually several ways to store 10:10:10:2 data, depending on channel order and whether the extra 2 bits are ignored or not.
The pixman library is probably the best reference for this.

**Comment by Oblomov** [comment]
Well, so much for that idea. RLE compression for these images actually results on a 30% _increase_ in size, so I'm guessing that path isn't a good idea, although I guess I could try some of the other compression schemes.

**Comment by urban-warrior** [comment]
You can save 10-bit images to TIFF by adding `-depth 10` to your command-line.  That would be 10:10:10 or 10:10:10:10 for images with an alpha channel.  Add `-compress zip` to the CLI to compress the image.  ImageMagick does not support images in the 10:10:10:2 format.

**Reply by Oblomov** [reply]
Thanks for the reply.  Since what I'm trying to achieve is to reduce the data _fed_ to IM (rather than the one produced by it) with as little work as possible on the producing side, I don't think IM's support for 10-bit TIFF would help me much in this case. So I guess what I'm doing now is the best I can do within the constraints.
